pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,696 @@
+	aksClusterConfigDynamic, err = e.dynamicClient.Namespace(namespace.GlobalNamespace).Update(context.TODO(), aksClusterConfigDynamic, v1.UpdateOptions{})
+	if err != nil {
+		return cluster, err
+	}
+
+	// AKS cluster and node pool statuses are not always immediately updated. This cause the AKSConfig to
+	// stay in "active" for a few seconds, causing the cluster to go back to "active".
+	timeout := time.NewTimer(10 * time.Second)
+	for {
+		select {
+		case event := <-w.ResultChan():
+			aksClusterConfigDynamic = event.Object.(*unstructured.Unstructured)
+			status, _ := aksClusterConfigDynamic.Object["status"].(map[string]interface{})
+			if status["phase"] == "active" {
+				continue
+			}
+
+			// this enqueue is necessary to ensure that the controller is reentered with the updating phase
+			e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// recordCAAndAPIEndpoint reads the AKSClusterConfig's secret once available. The CA cert and API endpoint are then copied to the cluster status.
+func (e *aksOperatorController) recordCAAndAPIEndpoint(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Factor:   2,
+		Jitter:   0,
+		Steps:    6,
+		Cap:      20 * time.Second,
+	}
+
+	var caSecret *corev1.Secret
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		var err error
+		caSecret, err = e.secretsCache.Get(namespace.GlobalNamespace, cluster.Name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return false, err
+			}
+			logrus.Infof("waiting for cluster [%s] data needed to generate service account token", cluster.Name)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return cluster, fmt.Errorf("failed waiting for cluster [%s] secret: %s", cluster.Name, err)
+	}
+
+	apiEndpoint := string(caSecret.Data["endpoint"])
+	caCert := string(caSecret.Data["ca"])
+	if cluster.Status.APIEndpoint == apiEndpoint && cluster.Status.CACert == caCert {
+		return cluster, nil
+	}
+
+	var currentCluster *mgmtv3.Cluster
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentCluster, err = e.clusterCache.Get(cluster.Name)
+		if err != nil {
+			return err
+		}
+		currentCluster.Status.APIEndpoint = apiEndpoint
+		currentCluster.Status.CACert = base64.StdEncoding.EncodeToString(caSecret.Data["ca"])
+		currentCluster, err = e.clusterClient.Update(currentCluster)
+		return err
+	})
+
+	return currentCluster, err
+}
+
+// generateAndSetServiceAccount uses the API endpoint and CA cert to generate a service account token. The token is then copied to the cluster status.
+func (e *aksOperatorController) generateAndSetServiceAccount(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
```

Comment:
Other than checking for the dnsError, there is no different between this and `generateSATokenWithPublicAPI`. There should be more of a difference. For example, this version should use the cluster dialer so the request is tunneled in the private case.

Reply:
This will be done for privateCluster